---
title: "Compare three different methods of extracting polyA sites in mouse sperm with the movAPA package"
author: "Tao Liu"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output:
  rmarkdown::html_document:
    toc:  true
    toc_depth: 6
    number_sections: true
    fig_width: 5
    fig_height: 4
    # df_print: kable
urlcolor: blue
fontsize: 11pt
vignette: >
  %\VignetteIndexEntry{movAPA_on_mouse_sperm_cells}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(knitr)
library(BiocStyle)
knitr::opts_chunk$set(echo = TRUE)
```
# Overview
Here we will use the movAPA toolkit to compare three different methods of extracting sites on the sperm data. These three methods are scAPAtrap, scAPA([Shulman et al, 2019](#1)), Sierra([Patrick, et al., 2020](#4)). We have used the scAPAtrap pipeline and refer to the pipeline of scAPA, Sierra to process the Sperm data, which can be downloaded by visiting this [link](http://bigbio.xmu.edu.cn/scAPAtrap/scPACdsData/).

For Sperm data, including early stage (spermatocytes, SC), intermediate stage (round spermatids, RS), and late stage (elongating spermatids, ES). For specific details, please refer to the article [Lukassen, et al., 2018](#3)

For detailed usage of [movAPA](#2) toolkit, please refer to [link](https://github.com/BMILAB/movAPA)

## Preparations
In this section, we will import the required R package and some required data. Please make sure you have downloaded the required data before running the code.
### Load R package
```{r message=FALSE}
library(ggplot2)
library(reshape2)
library(ggsci)
library(movAPA)
```

### Load Data
Load mouse reference genome, reference genome annotation, polyA site annotation.
```{r}
## Load mouse reference genome
library(BSgenome.Mmusculus.UCSC.mm10)
bsgenome <- BSgenome.Mmusculus.UCSC.mm10

## Load reference genome annotation
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdbmm10 <- parseGenomeAnnotation(TxDb.Mmusculus.UCSC.mm10.knownGene)

## Load polyA site annotation(polyAsite2.0)
load('C:/Users/admin/Desktop/refdata/mm10polyAsite2.rda')

## Load polyA site annotation(polyA DB3)
load('C:/Users/admin/Desktop/refdata/mm10polyAdb3.rda')
```

Load PolyA data calculated by three methods
```{r}
## scAPA
load("C:/Users/admin/Desktop/scPACdsData/scAPAds.rda")

## scAPAtrap
load("C:/Users/admin/Desktop/scPACdsData/scAPAtrapds.rda")

## sierra
load("C:/Users/admin/Desktop/scPACdsData/sierrads.rda")
```

## Basic statistics
In this section, we will do some basic statistics and graphing on the PAC extracted by the three methods.
### PAC number statistics
```{r}
## Total number of PACs
sum(scAPAtrapds@counts);sum(scAPAds@counts);sum(sierrads@counts)

## Average number of PACs
sum(scAPAtrapds@counts)/ncol(scAPAtrapds@counts);sum(scAPAds@counts)/ncol(scAPAtrapds@counts);sum(sierrads@counts)/ncol(sierrads@counts)
```

### PolyA regional distribution
```{r}
## scAPAtrap
table(scAPAtrapds@anno$ftr)

## scAPA
table(scAPAds@anno$ftr)

## sierra
table(sierrads@anno$ftr)
```

###  PAC's ATCG distribution map
```{r warning=FALSE}
## scAPAtrap
dsPlot = scAPAtrapds[scAPAtrapds@anno$ftr %in% c("3UTR", "CDS", "intron","intergenic")]
fafiles1 = faFromPACds(dsPlot, bsgenome, what = "updn", fapre = "scAPAtrap.400nt", up = -100, dn = 100, byGrp = "ftr", 
    chrCheck = FALSE)
plotATCGforFAfile(fafiles1, ofreq = TRUE, opdf = FALSE, refPos = 101, mergePlots = T, filepre = "scAPAtrap.400nt")
```

```{r warning=FALSE}
## sierra
dsPlot = sierrads[sierrads@anno$ftr %in% c("3UTR", "CDS", "intron","intergenic")]
fafiles2 = faFromPACds(dsPlot, bsgenome, what = "updn", fapre = "sierra.400nt", up = -100, dn = 100, byGrp = "ftr", 
    chrCheck = FALSE)
plotATCGforFAfile(fafiles2, ofreq = TRUE, opdf = FALSE, refPos = 101, mergePlots = T, filepre = "sierra.400nt")
```


```{r warning=FALSE}
## scAPA
dsPlot = scAPAds[scAPAds@anno$ftr %in% c("3UTR", "CDS", "intron","intergenic")]
fafiles3 = faFromPACds(dsPlot, bsgenome, what = "updn", fapre = "scAPA.400nt", up = -100, dn = 100, byGrp = "ftr", 
    chrCheck = FALSE)
plotATCGforFAfile(fafiles3, ofreq = TRUE, opdf = FALSE, refPos = 101, mergePlots = T, filepre = "scAPA.400nt")
```


```{r warning=FALSE}
## ploysite2.0
mm10polyAsite2<- annotatePAC(mm10polyAsite2,aGFF = txdbmm10)
mm10polyAsite2 <- ext3UTRPACds(mm10polyAsite2, ext3UTRlen=1000)
table(mm10polyAsite2@anno$ftr)

## polyAsite2.0
dsPlot = mm10polyAsite2[mm10polyAsite2@anno$ftr %in% c("3UTR", "CDS", "intron","intergenic")]
fafiles4 = faFromPACds(dsPlot, bsgenome, what = "updn", fapre = "mm10polyAsite2.400nt", up = -100, dn = 100, byGrp = "ftr", 
    chrCheck = FALSE)
plotATCGforFAfile(fafiles4, ofreq = TRUE, opdf = FALSE, refPos = 101, mergePlots = T, filepre = "mm10polyAsite2.400nt")
```


```{r warning=FALSE}
## ployAdb3
mm10polyAdb3<- annotatePAC(mm10polyAdb3,aGFF = txdbmm10)
mm10polyAdb3 <- ext3UTRPACds(mm10polyAdb3, ext3UTRlen=1000)
table(mm10polyAdb3@anno$ftr)
## polyAdb3 
dsPlot = mm10polyAdb3[mm10polyAdb3@anno$ftr %in% c("3UTR", "CDS", "intron")]
fafiles = faFromPACds(dsPlot, bsgenome, what = "updn", fapre = "mm10polyAdb3.400nt", up = -300, dn = 100, byGrp = "ftr", 
    chrCheck = FALSE)
plotATCGforFAfile(fafiles, ofreq = TRUE, opdf = FALSE, refPos = 301, mergePlots = T, filepre = "mm10polyAdb3.400nt")
```

## 3UTR regional PAC comparison
In this section, we will extract the polyA sites in the 3 UTR regions from the sites identified by the three methods,and discuss the accuracy of the three methods to identify the locus from the two aspects of the polyA signal of the site and the ratio associated with polyA annotationspolyA annotations.

```{r}
scAPAtrapds3UTR <- scAPAtrapds[scAPAtrapds@anno$ftr=='3UTR',]
scAPAds3UTR <- scAPAds[scAPAds@anno$ftr=='3UTR',]
sierrads3UTR <- sierrads[sierrads@anno$ftr == '3UTR',]
```
### Basic statistics
```{r}
## Three methods 3UTR area PAC number, and draw  histogram
PAC <- c(nrow(scAPAtrapds3UTR@counts),nrow(scAPAds3UTR@counts),nrow(sierrads3UTR@counts))
method <- factor(c('scAPAtrap','scAPA','sierra'),levels = c('scAPAtrap','scAPA','sierra'))

temp <- data.frame(PAC = PAC,method = method)

ggplot(temp) + aes(x = method, fill = method, colour = method,weight = PAC) + geom_bar() + scale_fill_nejm() + scale_color_nejm() + theme()
```

```{r}
## Number of genes
length(unique(scAPAtrapds3UTR@anno$gene));length(unique(scAPAds3UTR@anno$gene));length(unique(sierrads3UTR@anno$gene))

# Number of APA genes
scAPAtrapds3UTRh2 <- get3UTRAPAds(scAPAtrapds3UTR, sortPA = TRUE, choose2PA = NULL)
scAPAds3UTRh2 <- get3UTRAPAds(scAPAds3UTR, sortPA = TRUE, choose2PA = NULL)
sierrads3UTRh2 <- get3UTRAPAds(sierrads3UTR, sortPA = TRUE, choose2PA = NULL)

length(unique(scAPAtrapds3UTRh2@anno$gene));length(unique(scAPAds3UTRh2@anno$gene));length(unique(sierrads3UTRh2@anno$gene))
```
### Associate polyA annotations

Draw the proportion of polyA annotations associated with each method at different distances.

```{r message=F,results="hide"}
## polyA site2.0
scAPAtrap.map <- c()
scAPA.map <- c()
Sierra.map <- c()
for (d in 2:20){
  scAPAtrapds3UTR <- annotateByKnownPAC(scAPAtrapds3UTR, mm10polyAsite2, labels="Mm", d=5*d)
  scAPAtrap.map <- c(scAPAtrap.map,sum(scAPAtrapds3UTR@anno$Mm_ovp == 1))
  scAPAds3UTR <- annotateByKnownPAC(scAPAds3UTR, mm10polyAsite2, labels="Mm", d=5*d)
  scAPA.map <- c(scAPA.map,sum(scAPAds3UTR@anno$Mm_ovp == 1))
  sierrads3UTR <- annotateByKnownPAC(sierrads3UTR, mm10polyAsite2, labels="Mm", d=5*d)
  Sierra.map <- c(Sierra.map,sum(sierrads3UTR@anno$Mm_ovp == 1))
}
```

```{r warning=FALSE}
scAPAtrap.map <- round(scAPAtrap.map/length(scAPAtrapds3UTR),2)
scAPA.map <- round(scAPA.map/length(scAPAds3UTR),2)
Sierra.map <- round(Sierra.map/length(sierrads3UTR),2)
d <- c(2:20)*5
all.map <- data.frame(scAPAtrap = scAPAtrap.map,scAPA = scAPA.map, Sierra = Sierra.map,d=d)

all.map <- reshape2::melt(all.map, id = c("d"))

ggplot(all.map) + aes(x = d, y = value, fill = variable, colour = variable, shape = variable) + geom_point(size = 3L) + 
    geom_smooth(span = 0.33) + scale_fill_nejm() + scale_color_nejm() + annotate("text", x = 60, y = 0.9, label = "polyA site2.0", 
    family = "Arial", size = 10L) + labs(x = "Cutoff(distance)", y = "Relative to known PA(%)") + theme()
```

```{r message=F,results="hide"}
## polyA db3
scAPAtrap.map <- c()
scAPA.map <- c()
Sierra.map <- c()
for (d in 2:20){
  scAPAtrapds3UTR <- annotateByKnownPAC(scAPAtrapds3UTR, mm10polyAdb3, labels="Mm", d=5*d)
  scAPAtrap.map <- c(scAPAtrap.map,sum(scAPAtrapds3UTR@anno$Mm_ovp == 1))
  scAPAds3UTR <- annotateByKnownPAC(scAPAds3UTR, mm10polyAdb3, labels="Mm", d=5*d)
  scAPA.map <- c(scAPA.map,sum(scAPAds3UTR@anno$Mm_ovp == 1))
  sierrads3UTR <- annotateByKnownPAC(sierrads3UTR, mm10polyAdb3, labels="Mm", d=5*d)
  Sierra.map <- c(Sierra.map,sum(sierrads3UTR@anno$Mm_ovp == 1))
}
```

```{r warning=FALSE}
scAPAtrap.map <- round(scAPAtrap.map/length(scAPAtrapds3UTR),2)
scAPA.map <- round(scAPA.map/length(scAPAds3UTR),2)
Sierra.map <- round(Sierra.map/length(sierrads3UTR),2)
d <- c(2:20)*5
all.map <- data.frame(scAPAtrap = scAPAtrap.map,scAPA = scAPA.map, Sierra = Sierra.map,d=d)

all.map <- reshape2::melt(all.map, id = c("d"))

ggplot(all.map) + aes(x = d, y = value, fill = variable, colour = variable, shape = variable) + geom_point(size = 3L) + 
    geom_smooth(span = 0.33) + scale_fill_nejm() + scale_color_nejm() + annotate("text", x = 60, y = 0.9, label = "polyA DB3", 
    family = "Arial", size = 10L) + labs(x = "Cutoff(distance)", y = "Relative to known PA(%)") + theme()
```


### Distribution of the distance from the real site
In this part, we associate the sites identified by the three methods to the polyAsite2.0 annotation sites with a distance of d=100, and observe the distribution of the distances of the associated annotation sites when d=100.

```{r}
scAPAtrapds3UTR <- annotateByKnownPAC(scAPAtrapds3UTR, mm10polyAsite2, labels = "Mm", d = 100)
scAPAds3UTR <- annotateByKnownPAC(scAPAds3UTR, mm10polyAsite2, labels = "Mm", d = 100)
sierrads3UTR <- annotateByKnownPAC(sierrads3UTR , mm10polyAsite2, labels = "Mm", d = 100)
```
When d=100, the proportion of annotation sites on the association.

```{r}
sum(na.omit(scAPAtrapds3UTR@anno$Mm_dist))/sum(!is.na(scAPAtrapds3UTR@anno$Mm_dist));
sum(na.omit(scAPAds3UTR@anno$Mm_dist))/sum(!is.na(scAPAds3UTR@anno$Mm_dist));
sum(na.omit(sierrads3UTR@anno$Mm_dist))/sum(!is.na(sierrads3UTR@anno$Mm_dist))
```

Plot the distribution of distance.

```{r}
scAPA.dist <- as.data.frame(table(na.omit(scAPAds3UTR@anno$Mm_dist)))
scAPAtrap.dist <- as.data.frame(table(na.omit(scAPAtrapds3UTR@anno$Mm_dist)))
sierra.dist <- as.data.frame(table(na.omit(sierrads3UTR@anno$Mm_dist)))
scAPA.dist$Freq <- scAPA.dist$Freq/sum(scAPA.dist$Freq)
scAPAtrap.dist$Freq <- scAPAtrap.dist$Freq/sum(scAPAtrap.dist$Freq)
sierra.dist$Freq <- sierra.dist$Freq/sum(sierra.dist$Freq)
scAPA.dist$method <- rep("scAPA", length(scAPA.dist$Freq))
scAPAtrap.dist$method <- rep("scAPAtrap", length(scAPAtrap.dist$Freq))
sierra.dist$method <- rep("Sierra", length(sierra.dist$Freq))

all_dist <- rbind(scAPA.dist, scAPAtrap.dist, sierra.dist)
all_dist$Var1 <- as.numeric(all_dist$Var1)
all_dist$Var1 <- all_dist$Var1
all_dist$method <- factor(all_dist$method, levels = c("scAPAtrap", "Sierra", "scAPA"))
ggplot(all_dist) + aes(x = Var1, y = Freq, fill = method, colour = method) + geom_line(size = 1L) + scale_fill_nejm() + 
    scale_color_nejm() + labs(x = "Cutoff(distance)", y = "% of PA", fill = "method", color = "method") + theme()
```

```{r include=FALSE}
## scAPAtrap
scAPAtrapds3UTR <- annotateByKnownPAC(scAPAtrapds3UTR, mm10polyAdb3, labels = "Mm", d = 100)
index2 <- which(scAPAtrapds3UTR@anno$Mm_ovp == 1)
scAPAtrapds3UTR <- annotateByKnownPAC(scAPAtrapds3UTR, mm10polyAsite2, labels = "Mm", d = 100)
index3 <- which(scAPAtrapds3UTR@anno$Mm_ovp == 1)
nrow(scAPAtrapds3UTR@counts)-length(union(index2,index3))
```

```{r include=FALSE}
## scAPA
scAPAds3UTR <- annotateByKnownPAC(scAPAds3UTR, mm10polyAdb3, labels = "Mm", d = 100)
index2 <- which(scAPAds3UTR@anno$Mm_ovp == 1)
scAPAds3UTR <- annotateByKnownPAC(scAPAds3UTR, mm10polyAsite2, labels = "Mm", d = 100)
index3 <- which(scAPAds3UTR@anno$Mm_ovp == 1)
nrow(scAPAds3UTR@counts)-length(union(index2,index3))
```

```{r include=FALSE}
sierrads3UTR <- annotateByKnownPAC(sierrads3UTR, mm10polyAdb3, labels = "Mm", d = 100)
index2 <- which(sierrads3UTR@anno$Mm_ovp == 1)
sierra3UTR <- annotateByKnownPAC(sierrads3UTR, mm10polyAsite2, labels = "Mm", d = 100)
index3 <- which(sierra3UTR@anno$Mm_ovp == 1)
nrow(sierra3UTR@counts)-length(union(index2,index3))
```
### Calculation of polyA signal

```{r}
## Imported mouse polyA signal
load('C:/Users/admin/Desktop/refdata/mm10Signal.rda')
## Set priority
priority <- c(1, 2, rep(3, 16))
```

```{r}
## scAPAtrap
scAPAtrapds3UTR <- annotateByPAS(pacds = scAPAtrapds3UTR, bsgenome, grams = mm10Signal, priority = priority, from = -60, to = 10, 
    label = "Mm", chrCheck = TRUE)
table(scAPAtrapds3UTR@anno$Mm_gram)
sum(!is.na(scAPAtrapds3UTR@anno$Mm_gram))/length(scAPAtrapds3UTR)  
```

```{r}
## scAPA
scAPAds3UTR <- annotateByPAS(pacds = scAPAds3UTR, bsgenome, grams = mm10Signal, priority = priority, from = -60, to = 10, 
    label = "Mm", chrCheck = TRUE)
table(scAPAds3UTR@anno$Mm_gram)
sum(!is.na(scAPAds3UTR@anno$Mm_gram))/length(scAPAds3UTR)
```

```{r}
## sierra
sierrads3UTR <- annotateByPAS(pacds = sierrads3UTR, bsgenome, grams = mm10Signal, priority = priority, from = -60, to = 10, 
    label = "Mm", chrCheck = TRUE)
table(sierrads3UTR@anno$Mm_gram)
sum(!is.na(sierrads3UTR@anno$Mm_gram))/length(sierrads3UTR)
```

```{r include=FALSE}
## mm1opolysite2.0
mm10polyAsite23UTR <-mm10polyAsite2[mm10polyAsite2@anno$ftr == '3UTR',]
mm10polyAsite23UTR<- annotateByPAS(pacds = mm10polyAsite23UTR, bsgenome, grams = mm10Signal, priority = priority, from = -60, to = 10, 
    label = "Mm", chrCheck = TRUE)
table(mm10polyAsite23UTR@anno$Mm_gram)
sum(!is.na(mm10polyAsite23UTR@anno$Mm_gram))/length(mm10polyAsite23UTR) # 0.6712074
```

Draw the proportion of each method signal

```{R}
scAPA.signal <- scAPAds3UTR@anno$Mm_gram[!is.na(scAPAds3UTR@anno$Mm_gram)]
scAPA.signal[!scAPA.signal %in% c("AATAAA", "ATTAAA")] <- "1nt-variants"
scAPA.signal <- as.data.frame(table(scAPA.signal))
scAPA.signal$Freq <- scAPA.signal$Freq/length(scAPAds3UTR@anno$Mm_gram)
scAPA.signal$methed <- rep("scAPA", length(scAPA.signal$Freq))
colnames(scAPA.signal) <- c("signal", "freq", "method")

scAPAtrap.signal <- scAPAtrapds3UTR@anno$Mm_gram[!is.na(scAPAtrapds3UTR@anno$Mm_gram)]
scAPAtrap.signal[!scAPAtrap.signal %in% c("AATAAA", "ATTAAA")] <- "1nt-variants"
scAPAtrap.signal <- as.data.frame(table(scAPAtrap.signal))
scAPAtrap.signal$Freq <- scAPAtrap.signal$Freq/length(scAPAtrapds3UTR@anno$Mm_gram)
scAPAtrap.signal$methed <- rep("scAPAtrap", length(scAPAtrap.signal$Freq))
colnames(scAPAtrap.signal) <- c("signal", "freq", "method")

sierra.signal <- sierrads3UTR@anno$Mm_gram[!is.na(sierrads3UTR@anno$Mm_gram)]
sierra.signal[!sierra.signal %in% c("AATAAA", "ATTAAA")] <- "1nt-variants"
sierra.signal <- as.data.frame(table(sierra.signal))
sierra.signal$Freq <- sierra.signal$Freq/length(sierrads3UTR@anno$Mm_gram)
sierra.signal$methed <- rep("Sierra", length(sierra.signal$Freq))
colnames(sierra.signal) <- c("signal", "freq", "method")

all_signal <- rbind(scAPAtrap.signal, scAPA.signal, sierra.signal)
all_signal$signal = factor(all_signal$signal, levels = c("1nt-variants", "ATTAAA", "AATAAA"))
all_signal$method <- factor(all_signal$method, levels = c("scAPAtrap", "scAPA", "Sierra",'Bulk'))
ggplot(all_signal) + aes(x = method, fill = signal, colour = signal, weight = freq) + geom_bar(position = "stack") + scale_fill_nejm() + 
    scale_color_nejm() + labs(y = "% of signal", fill = "method", color = "method") + theme()
```

### PAC's overlaping

```{r results="hide",message=FALSE}
scAPAtrapds3UTR.G<- with(scAPAtrapds3UTR@anno, GRanges(seqnames = chr, ranges = IRanges(start = start, end = end), strand = strand))
scAPAds3UTR.G <- with(scAPAds3UTR@anno, GRanges(seqnames = chr, ranges = IRanges(start = start, end = end), strand = strand))
sierrads3UTR.G <- with(sierrads3UTR@anno, GRanges(seqnames = chr, ranges = IRanges(start = start, end = end), strand = strand))

ov <- ChIPpeakAnno::findOverlapsOfPeaks(scAPAtrapds3UTR.G, scAPAds3UTR.G, sierrads3UTR.G , ignore.strand = FALSE)
ChIPpeakAnno::makeVennDiagram(ov, fill = c("#FF69B4", "#32CD32", "#FFFF00"))
```

## Identification of low expression PAC

In this section, we will discuss the effects of the three methods on low-expressed PAC recognition.
### Basic statistics
Here we think that the total expression level is less than 30 as a low expression level, so cutoff is set to 30.
```{r}
cutoff <- 30
scAPAtrapds3UTR.low <- scAPAtrapds3UTR[rowSums(scAPAtrapds3UTR@counts) <= cutoff, ]
scAPAds3UTR.low <- scAPAds3UTR[rowSums(scAPAds3UTR@counts) <= cutoff, ]
sierrads3UTR.low <- sierrads3UTR[rowSums(sierrads3UTR@counts) <= cutoff, ]
```

Number of low-expressed PACs by each method
```{r}
length(scAPAtrapds3UTR.low);length(scAPAds3UTR.low);length(sierrads3UTR.low)
```

Number of PACs with low expression and polyA signal

```{r}
sum(!is.na(scAPAtrapds3UTR.low@anno$Mm_gram));sum(!is.na(scAPAds3UTR.low@anno$Mm_gram));sum(!is.na(sierrads3UTR.low@anno$Mm_gram))
```

Ratio of PACs with low expression and polyA signal
```{r}
sum(!is.na(scAPAtrapds3UTR.low@anno$Mm_gram))/length(scAPAtrapds3UTR.low);sum(!is.na(scAPAds3UTR.low@anno$Mm_gram))/length(scAPAds3UTR.low);sum(!is.na(sierrads3UTR.low@anno$Mm_gram))/length(sierrads3UTR.low);
```


Number of polyA annotations associated with low expression PAC
```{r}
sum(scAPAtrapds3UTR.low@anno$Mm_ovp);sum(scAPAds3UTR.low@anno$Mm_ovp);sum(sierrads3UTR.low@anno$Mm_ovp)
```


Ratio of polyA annotations associated with low expression PAC
```{r}
sum(scAPAtrapds3UTR.low@anno$Mm_ovp)/length(scAPAtrapds3UTR.low);sum(scAPAds3UTR.low@anno$Mm_ovp)/length(scAPAds3UTR.low);sum(sierrads3UTR.low@anno$Mm_ovp)/length(sierrads3UTR.low);
```

### Overlapping of low expression PAC
```{r results="hide",message=FALSE}
scAPAtrap.g_low <- with(scAPAtrapds3UTR.low@anno, GRanges(seqnames = chr, ranges = IRanges(start = start, end = end), strand = strand))
scAPA.g_low <- with(scAPAds3UTR.low@anno, GRanges(seqnames = chr, ranges = IRanges(start = start, end = end), strand = strand))
sierra.g_low <- with(sierrads3UTR.low@anno, GRanges(seqnames = chr, ranges = IRanges(start = start, end = end), strand = strand))

ov1 <- ChIPpeakAnno::findOverlapsOfPeaks(scAPAtrap.g_low , scAPA.g_low,sierra.g_low, ignore.strand = FALSE)
ChIPpeakAnno::makeVennDiagram(ov1, fill = c("#FF69B4", "#32CD32", "#FFFF00"))
```

### motif and ATCG distribution map drawing
```{r warning=FALSE}
## ATCG distribution map
fafiles = faFromPACds(scAPAtrapds3UTR.low, bsgenome, what = "updn", fapre = "scAPAtrap3UTR.low", up = -300, dn = 100, chrCheck = FALSE)
plotATCGforFAfile(fafiles, ofreq = TRUE, opdf = F, refPos = 301, mergePlots = F, filepre = "scAPAtrap3UTR.low")
```

```{r message=FALSE,results="hide"}
## motif
pas = scAPAtrapds3UTR.low@anno$Mm_gram[!is.na(scAPAtrapds3UTR.low@anno$Mm_gram)]
plotSeqLogo(pas)
```

### The effect of low expression PAC on the number of genes
Changes in the number of genes before and after the introduction of low expression PAC

```{r}
scAPAtrapds3UTR.upper <- scAPAtrapds3UTR[rowSums(scAPAtrapds3UTR@counts) > cutoff, ]
temp1 <- dplyr::group_by(scAPAtrapds3UTR.upper@anno, gene) %>% summarise(count = n())
temp2 <- dplyr::group_by(scAPAtrapds3UTR@anno, gene) %>% summarise(count = n())

temp1_num <- c()
temp2_num <- c()
for (i in c(1:5)) {
    if (i < 5) {
        temp1_num <- c(temp1_num, sum(temp1$count == i))
        temp2_num <- c(temp2_num, sum(temp2$count == i))
    } else {
        temp1_num <- c(temp1_num, sum(temp1$count >= i))
        temp2_num <- c(temp2_num, sum(temp2$count >= i))
    }
}

nPAC <- paste0(c("=", "=", "=", "=", ">="), c(1:5))
lowplot <- data.frame(before = temp1_num, after = temp2_num, nPAC = nPAC)
lowplot

lowplot <- reshape2::melt(lowplot, id = c("nPAC"))
ggplot(lowplot) + aes(x = nPAC, fill = variable, colour = variable, weight = value) + geom_bar(position = "dodge") + scale_fill_nejm() + 
    scale_color_nejm() + labs(y = "The number of genes") + theme()
```


```{r include=FALSE}
subset(scAPAtrapds3UTR@anno,chr == 'chr4' & start >= 32773900 & end <= 32775217)
rowSums(scAPAtrapds3UTR@counts[scAPAtrapds3UTR@anno$gene == '100019',])
subset(mm10polyAdb3@anno,chr == 'chr4' & start > 32775216-100 & end < 32775216 + 100)
subset(mm10polyAsite2@anno,chr == 'chr4' & start > 32775216-100 & end < 32775216 + 100)
```

# Session Information
The session information records the versions of all the packages used in the generation of the present document.
```{r}
sessionInfo()
```

# References{#refer}
<span id="1">[1] Shulman, E.D. and Elkon, R. Cell-type-specific analysis of alternative polyadenylation using single-cell transcriptomics data. Nucleic Acids Res 2019;47(19):10027-10039.</span>

<span id="2">[2] movAPA: Modeling and visualization of dynamics of alternative polyadenylation across biological samples (under review).</span>

<span id="3">[3] Lukassen, S., et al. (2018) Single-cell RNA sequencing of adult mouse testes, Scientific Data, 5, 180192.</span>

<span id="4">[4] Sierra: discovery of differential transcript usage from polyA-captured single-cell RNA-seq data </span>
